### PR TITLE
CMake: add a safeguard for accidential /source source directory

### PR DESCRIPTION
**Pull request by tamiko**
This happens remarkably often. Let's improve the error message:
```
  Wrong source directory specified.

  You pointed cmake to the subdirectory "source" instead of the base
  directory.  Make sure that the path in the cmake invocation does not end in
  source.  If you have created a "build" subdirectory, the cmake invocation
  should simply be of the form:

    $ cmake  [...options...]  /..

  (path not ending in ".../source")
```

### CODE REVIEW

**Comment by tamiko** [comment]
@jppelteret I hope you appreciate this pull request ;-)

**Comment by jppelteret** [comment]
> I hope you appreciate this pull request ;-)

I do indeed, and I'm sure our first-time users will too. Thanks a lot!

**Comment by bangerth** [comment]
On 06/03/2018 02:35 PM, Matthias Maier wrote:
> |You pointed cmake to the subdirectory "source" instead of the base directory. 
> Make sure that the path in the cmake invocation does not end in source. If you 
> have created a "build" subdirectory, the cmake invocation should simply be of 
> the form:|

I think that's good in principle, but the wording is awkward. That's because 
both "source directory" and "base directory" are ambiguous terms -- in the 
former case, it may be `deal.II/` or `deal.II/source/`; in the latter case it 
may be `deal.II/` or `deal.II/source/base/` or even 
`deal.II/include/deal.II/base`.

Maybe rephrase this as follows:


   When telling cmake about the directory in which the deal.II source files are
   located, you pointed to a subdirectory of the form "deal.II/source/" or
   "dealiii/source/" or similar, instead of the top level directory in which
   both the .cc and .h files (and many other files) are located.
   Make sure that instead you use the top-level path in the cmake invocation,
   i.e., the name of the given directory can not end in
   "source".  If you have created a "build" subdirectory as a subdirectory of
   the source directory, then the cmake invocation
   should simply be of the form:

File: source/CMakeLists.txt [hunk 1]
```diff
@@ -15,6 +15,22 @@
...
+#
+IF("${DEAL_II_PACKAGE_NAME}" STREQUAL  "")
+  MESSAGE(FATAL_ERROR
+    "\nWrong source directory specified.\n"
+    "You pointed cmake to the subdirectory \"source\" instead of the base "
+    "directory (the one containing the file \"README.md\"). Make sure that "
+    "the path in the cmake invocation does not end in \".../source\". If you "
+    "have created a \"build\" subdirectory, the cmake invocation should simply "
+    "be of the form:\n"
+    "  $ cmake  [...options...]  /..\n"
```
Comment (bangerth) [inline source/CMakeLists.txt]: It needs to be `../`, not `/..`, I think.

Comment (tamiko) [reply source/CMakeLists.txt]: fixed.

File: source/CMakeLists.txt [hunk 2]
```diff
@@ -26,7 +26,7 @@ IF("${DEAL_II_PACKAGE_NAME}" STREQUAL  "")
     "the path in the cmake invocation does not end in \".../source\". If you "
```
Comment (masterleinad) [inline source/CMakeLists.txt]: This should also be `../source>` instead of `.../source`, I assume.

File: source/CMakeLists.txt [hunk 3]
```diff
@@ -26,7 +26,7 @@ IF("${DEAL_II_PACKAGE_NAME}" STREQUAL  "")
...
     "have created a \"build\" subdirectory, the cmake invocation should simply "
     "be of the form:\n"
-    "  $ cmake  [...options...]  /..\n"
+    "  $ cmake  [...options...]  ../\n"
     "(path not ending in \".../source\")"
```
Comment (masterleinad) [inline source/CMakeLists.txt]: Same as above.

Comment (tamiko) [reply source/CMakeLists.txt]: no, this should be an ellipsis "..." and not previous directory "..".

Comment (masterleinad) [reply source/CMakeLists.txt]: OK

**Comment by bangerth** [comment]
Yes, the ellipsis was correct. Thanks for rewording this!
